mongodb select * on view returns error CommandNotSupportedOnView

Any Idea why a select * on a mongodb view results in this error:
```
SQL-Fehler [65536]: Query failed (#20210403_015848_00023_3siug): com.mongodb.MongoCommandException: 
Command failed with error 166 (CommandNotSupportedOnView): 'Namespace enioSites.enrichedSites is a view,
 not a collection' on server host.docker.internal:27117.
 The full response is 
```
```json
{
  "operationTime": {
    "$timestamp": {
      "t": 1617453002,
      "i": 1
    }
  },
  "ok": 0,
  "errmsg": "Namespace enioSites.enrichedSites is a view, not a collection",
  "code": 166,
  "codeName": "CommandNotSupportedOnView",
  "$clusterTime": {
    "clusterTime": {
      "$timestamp": {
        "t": 1617453002,
        "i": 1
      }
    },
    "signature": {
      "hash": {
        "$binary": "AAAAAAAAAAAAAAAAAAAAAAAAAAA=",
        "$type": "00"
      },
      "keyId": {
        "$numberLong": "0"
      }
    }
  }
}
```

Thanks
Ognian

OK, since I cannot share my data I used sample data to reproduce the error.

I'm using MongoDB 4.2.7 Community

Go to https://github.com/SouthbankSoftware/dbkoda-data and Download as zip
unzip and go to OrdersExample directory
`mongorestore --uri mongodb://localhost:27117/test dump`
you have now an OrderExample database

use the latest MongoDB Compass (Version 1.26.1-beta.6 (1.26.1-beta.6)) to add the following mongodb view:
1. go to the orders collection and go to aggregations
2. click on the drop down arrow next to the + and choose `new pipeline from text` 
<img width="1277" alt="image" src="https://user-images.githubusercontent.com/1231174/113571166-22aad500-9616-11eb-80b5-5dd1b689459f.png">

4. and add
```
[{$lookup: {
  from: 'customers',
  localField: 'customerId',
  foreignField: '_id',
  as: 'customer'
}}]
```
3 click on the drop down arrow next to the green save button and choose create view, name it 'lookupCustomer'
<img width="508" alt="image" src="https://user-images.githubusercontent.com/1231174/113571202-38b89580-9616-11eb-85d3-08b5593bb961.png">

Now you have a 'lookupCustomer' View, and it appears in Compass on the left side with the other OrdersExample collections. By clicking on `lookupCustomer` check that we have indeed the new field customer with one Object
<img width="1280" alt="image" src="https://user-images.githubusercontent.com/1231174/113571052-e37c8400-9615-11eb-905d-39039a67e145.png">

Now go to i.e dbeaver and see the error:
<img width="1920" alt="image" src="https://user-images.githubusercontent.com/1231174/113571410-a49afe00-9616-11eb-9c37-abeb8bc13134.png">

As you can see Im using the newest trino version: 354
<img width="1344" alt="image" src="https://user-images.githubusercontent.com/1231174/113571522-d3b16f80-9616-11eb-879a-454371ef770a.png">

query:
```
SELECT TABLE_CAT, TABLE_SCHEM, TABLE_NAME, COLUMN_NAME, DATA_TYPE,
  TYPE_NAME, COLUMN_SIZE, BUFFER_LENGTH, DECIMAL_DIGITS, NUM_PREC_RADIX,
  NULLABLE, REMARKS, COLUMN_DEF, SQL_DATA_TYPE, SQL_DATETIME_SUB,
  CHAR_OCTET_LENGTH, ORDINAL_POSITION, IS_NULLABLE,
  SCOPE_CATALOG, SCOPE_SCHEMA, SCOPE_TABLE,
  SOURCE_DATA_TYPE, IS_AUTOINCREMENT, IS_GENERATEDCOLUMN
FROM system.jdbc.columns
WHERE TABLE_CAT = 'mongodb' AND TABLE_SCHEM LIKE 'orderexample' ESCAPE '\' AND TABLE_NAME LIKE 'lookupcustomer' ESCAPE '\' AND COLUMN_NAME LIKE '%' ESCAPE '\'
ORDER BY TABLE_CAT, TABLE_SCHEM, TABLE_NAME, ORDINAL_POSITION
```
stacktrace:
```
com.google.common.util.concurrent.UncheckedExecutionException: com.mongodb.MongoCommandException: Command failed with error 166 (CommandNotSupportedOnView): 'Namespace OrderExample.lookupCustomer is a view, not a collection' on server host.docker.internal:27117. The full response is { "operationTime" : { "$timestamp" : { "t" : 1617622153, "i" : 1 } }, "ok" : 0.0, "errmsg" : "Namespace OrderExample.lookupCustomer is a view, not a collection", "code" : 166, "codeName" : "CommandNotSupportedOnView", "$clusterTime" : { "clusterTime" : { "$timestamp" : { "t" : 1617622153, "i" : 1 } }, "signature" : { "hash" : { "$binary" : "AAAAAAAAAAAAAAAAAAAAAAAAAAA=", "$type" : "00" }, "keyId" : { "$numberLong" : "0" } } } }
	at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2051)
	at com.google.common.cache.LocalCache.get(LocalCache.java:3951)
	at com.google.common.cache.LocalCache.getOrLoad(LocalCache.java:3974)
	at com.google.common.cache.LocalCache$LocalLoadingCache.get(LocalCache.java:4935)
	at com.google.common.cache.LocalCache$LocalLoadingCache.getUnchecked(LocalCache.java:4941)
	at io.trino.plugin.mongodb.MongoSession.getTable(MongoSession.java:167)
	at io.trino.plugin.mongodb.MongoMetadata.getTableMetadata(MongoMetadata.java:315)
	at io.trino.plugin.mongodb.MongoMetadata.listTableColumns(MongoMetadata.java:133)
	at io.trino.metadata.MetadataManager.listTableColumns(MetadataManager.java:608)
	at io.trino.metadata.MetadataListing.listTableColumns(MetadataListing.java:135)
	at io.trino.connector.system.jdbc.ColumnJdbcTable.cursor(ColumnJdbcTable.java:257)
	at io.trino.connector.system.SystemPageSourceProvider$1.cursor(SystemPageSourceProvider.java:128)
	at io.trino.split.MappedRecordSet.cursor(MappedRecordSet.java:53)
	at io.trino.spi.connector.RecordPageSource.<init>(RecordPageSource.java:37)
	at io.trino.connector.system.SystemPageSourceProvider.createPageSource(SystemPageSourceProvider.java:107)
	at io.trino.spi.connector.ConnectorPageSourceProvider.createPageSource(ConnectorPageSourceProvider.java:68)
	at io.trino.split.PageSourceManager.createPageSource(PageSourceManager.java:64)
	at io.trino.operator.ScanFilterAndProjectOperator$SplitToPages.process(ScanFilterAndProjectOperator.java:254)
	at io.trino.operator.ScanFilterAndProjectOperator$SplitToPages.process(ScanFilterAndProjectOperator.java:182)
	at io.trino.operator.WorkProcessorUtils$3.process(WorkProcessorUtils.java:319)
	at io.trino.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:372)
	at io.trino.operator.WorkProcessorUtils$3.process(WorkProcessorUtils.java:306)
	at io.trino.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:372)
	at io.trino.operator.WorkProcessorUtils$3.process(WorkProcessorUtils.java:306)
	at io.trino.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:372)
	at io.trino.operator.WorkProcessorUtils.getNextState(WorkProcessorUtils.java:221)
	at io.trino.operator.WorkProcessorUtils.lambda$processStateMonitor$2(WorkProcessorUtils.java:200)
	at io.trino.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:372)
	at io.trino.operator.WorkProcessorUtils.getNextState(WorkProcessorUtils.java:221)
	at io.trino.operator.WorkProcessorUtils.lambda$finishWhen$3(WorkProcessorUtils.java:215)
	at io.trino.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:372)
	at io.trino.operator.WorkProcessorSourceOperatorAdapter.getOutput(WorkProcessorSourceOperatorAdapter.java:149)
	at io.trino.operator.Driver.processInternal(Driver.java:387)
	at io.trino.operator.Driver.lambda$processFor$9(Driver.java:291)
	at io.trino.operator.Driver.tryWithLock(Driver.java:683)
	at io.trino.operator.Driver.processFor(Driver.java:284)
	at io.trino.execution.SqlTaskExecution$DriverSplitRunner.processFor(SqlTaskExecution.java:1075)
	at io.trino.execution.executor.PrioritizedSplitRunner.process(PrioritizedSplitRunner.java:163)
	at io.trino.execution.executor.TaskExecutor$TaskRunner.run(TaskExecutor.java:484)
	at io.trino.$gen.Trino_354____20210404_135930_2.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: com.mongodb.MongoCommandException: Command failed with error 166 (CommandNotSupportedOnView): 'Namespace OrderExample.lookupCustomer is a view, not a collection' on server host.docker.internal:27117. The full response is { "operationTime" : { "$timestamp" : { "t" : 1617622153, "i" : 1 } }, "ok" : 0.0, "errmsg" : "Namespace OrderExample.lookupCustomer is a view, not a collection", "code" : 166, "codeName" : "CommandNotSupportedOnView", "$clusterTime" : { "clusterTime" : { "$timestamp" : { "t" : 1617622153, "i" : 1 } }, "signature" : { "hash" : { "$binary" : "AAAAAAAAAAAAAAAAAAAAAAAAAAA=", "$type" : "00" }, "keyId" : { "$numberLong" : "0" } } } }
	at com.mongodb.internal.connection.ProtocolHelper.getCommandFailureException(ProtocolHelper.java:179)
	at com.mongodb.internal.connection.InternalStreamConnection.receiveCommandMessageResponse(InternalStreamConnection.java:293)
	at com.mongodb.internal.connection.InternalStreamConnection.sendAndReceive(InternalStreamConnection.java:255)
	at com.mongodb.internal.connection.UsageTrackingInternalConnection.sendAndReceive(UsageTrackingInternalConnection.java:99)
	at com.mongodb.internal.connection.DefaultConnectionPool$PooledConnection.sendAndReceive(DefaultConnectionPool.java:444)
	at com.mongodb.internal.connection.CommandProtocolImpl.execute(CommandProtocolImpl.java:72)
	at com.mongodb.internal.connection.DefaultServer$DefaultServerProtocolExecutor.execute(DefaultServer.java:200)
	at com.mongodb.internal.connection.DefaultServerConnection.executeProtocol(DefaultServerConnection.java:269)
	at com.mongodb.internal.connection.DefaultServerConnection.command(DefaultServerConnection.java:131)
	at com.mongodb.internal.connection.DefaultServerConnection.command(DefaultServerConnection.java:123)
	at com.mongodb.operation.CommandOperationHelper.executeWrappedCommandProtocol(CommandOperationHelper.java:242)
	at com.mongodb.operation.CommandOperationHelper.executeWrappedCommandProtocol(CommandOperationHelper.java:233)
	at com.mongodb.operation.CommandOperationHelper.executeWrappedCommandProtocol(CommandOperationHelper.java:136)
	at com.mongodb.operation.ListIndexesOperation$1.call(ListIndexesOperation.java:143)
	at com.mongodb.operation.ListIndexesOperation$1.call(ListIndexesOperation.java:138)
	at com.mongodb.operation.OperationHelper.withConnectionSource(OperationHelper.java:457)
	at com.mongodb.operation.OperationHelper.withConnection(OperationHelper.java:401)
	at com.mongodb.operation.ListIndexesOperation.execute(ListIndexesOperation.java:138)
	at com.mongodb.operation.ListIndexesOperation.execute(ListIndexesOperation.java:69)
	at com.mongodb.client.internal.MongoClientDelegate$DelegateOperationExecutor.execute(MongoClientDelegate.java:179)
	at com.mongodb.client.internal.MongoIterableImpl.execute(MongoIterableImpl.java:132)
	at com.mongodb.client.internal.MongoIterableImpl.iterator(MongoIterableImpl.java:86)
	at io.trino.plugin.mongodb.MongoIndex.parse(MongoIndex.java:36)
	at io.trino.plugin.mongodb.MongoSession.getIndexes(MongoSession.java:266)
	at io.trino.plugin.mongodb.MongoSession.loadTableSchema(MongoSession.java:226)
	at com.google.common.cache.CacheLoader$FunctionToCacheLoader.load(CacheLoader.java:165)
	at com.google.common.cache.LocalCache$LoadingValueReference.loadFuture(LocalCache.java:3529)
	at com.google.common.cache.LocalCache$Segment.loadSync(LocalCache.java:2278)
	at com.google.common.cache.LocalCache$Segment.lockedGetOrLoad(LocalCache.java:2155)
	at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2045)
	... 42 more
```

--- a/plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/MongoSession.java
+++ b/plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/MongoSession.java
@@ -188,7 +188,10 @@ public class MongoSession
 
     public void addColumn(SchemaTableName schemaTableName, ColumnMetadata columnMetadata)
     {
-        Document metadata = getTableMetadata(schemaTableName);
+        String schemaName = toRemoteSchemaName(schemaTableName.getSchemaName());
+        String tableName = toRemoteTableName(schemaName, schemaTableName.getTableName());
+
+        Document metadata = getTableMetadata(schemaName, tableName);
 
         List<Document> columns = new ArrayList<>(getColumnMetadata(metadata));
 
@@ -197,9 +200,6 @@ public class MongoSession
         newColumn.append(FIELDS_TYPE_KEY, columnMetadata.getType().getTypeSignature().toString());
         newColumn.append(FIELDS_HIDDEN_KEY, false);
         columns.add(newColumn);
-
-        String schemaName = toRemoteSchemaName(schemaTableName.getSchemaName());
-        String tableName = toRemoteTableName(schemaName, schemaTableName.getTableName());
 
         metadata.append(FIELDS_KEY, columns);
 
@@ -210,10 +210,13 @@ public class MongoSession
         tableCache.invalidate(schemaTableName);
     }
 
-    private MongoTable loadTableSchema(SchemaTableName tableName)
+    private MongoTable loadTableSchema(SchemaTableName schemaTableName)
             throws TableNotFoundException
     {
-        Document tableMeta = getTableMetadata(tableName);
+        String schemaName = toRemoteSchemaName(schemaTableName.getSchemaName());
+        String tableName = toRemoteTableName(schemaName, schemaTableName.getTableName());
+
+        Document tableMeta = getTableMetadata(schemaName, tableName);
 
         ImmutableList.Builder<MongoColumnHandle> columnHandles = ImmutableList.builder();
 
@@ -222,8 +225,8 @@ public class MongoSession
             columnHandles.add(columnHandle);
         }
 
-        MongoTableHandle tableHandle = new MongoTableHandle(tableName);
-        return new MongoTable(tableHandle, columnHandles.build(), getIndexes(tableName));
+        MongoTableHandle tableHandle = new MongoTableHandle(schemaTableName);
+        return new MongoTable(tableHandle, columnHandles.build(), getIndexes(schemaName, tableName));
     }
 
     private MongoColumnHandle buildColumnHandle(Document columnMeta)
@@ -258,12 +261,13 @@ public class MongoSession
         return client.getDatabase(schemaName).getCollection(tableName);
     }
 
-    public List<MongoIndex> getIndexes(SchemaTableName tableName)
+    public List<MongoIndex> getIndexes(String schemaName, String tableName)
     {
-        if (isView(tableName)) {
+        if (isView(schemaName, tableName)) {
             return ImmutableList.of();
         }
-        return MongoIndex.parse(getCollection(tableName).listIndexes());
+        MongoCollection<Document> collection = client.getDatabase(schemaName).getCollection(tableName);
+        return MongoIndex.parse(collection.listIndexes());
     }
 
     public MongoCursor<Document> execute(MongoTableHandle tableHandle, List<MongoColumnHandle> columns)
@@ -418,12 +422,9 @@ public class MongoSession
     }
 
     // Internal Schema management
-    private Document getTableMetadata(SchemaTableName schemaTableName)
+    private Document getTableMetadata(String schemaName, String tableName)
             throws TableNotFoundException
     {
-        String schemaName = toRemoteSchemaName(schemaTableName.getSchemaName());
-        String tableName = toRemoteTableName(schemaName, schemaTableName.getTableName());
-
         MongoDatabase db = client.getDatabase(schemaName);
         MongoCollection<Document> schema = db.getCollection(schemaCollection);
 
@@ -432,7 +433,7 @@ public class MongoSession
 
         if (doc == null) {
             if (!collectionExists(db, tableName)) {
-                throw new TableNotFoundException(schemaTableName);
+                throw new TableNotFoundException(new SchemaTableName(schemaName, tableName), format("Table '%s.%s' not found", schemaName, tableName), null);
             }
             else {
                 Document metadata = new Document(TABLE_NAME_KEY, tableName);
@@ -640,14 +641,14 @@ public class MongoSession
         return tableName;
     }
 
-    private boolean isView(SchemaTableName tableName)
+    private boolean isView(String schemaName, String tableName)
     {
         Document listCollectionsCommand = new Document(new ImmutableMap.Builder<String, Object>()
                 .put("listCollections", 1.0)
-                .put("filter", documentOf("name", tableName.getTableName()))
+                .put("filter", documentOf("name", tableName))
                 .put("nameOnly", true)
                 .build());
-        Document cursor = client.getDatabase(tableName.getSchemaName()).runCommand(listCollectionsCommand).get("cursor", Document.class);
+        Document cursor = client.getDatabase(schemaName).runCommand(listCollectionsCommand).get("cursor", Document.class);
         List<Document> firstBatch = cursor.get("firstBatch", List.class);
         if (firstBatch.isEmpty()) {
             return false;

--- a/plugin/trino-mongodb/src/test/java/io/trino/plugin/mongodb/TestMongoIntegrationSmokeTest.java
+++ b/plugin/trino-mongodb/src/test/java/io/trino/plugin/mongodb/TestMongoIntegrationSmokeTest.java
@@ -326,6 +326,34 @@ public class TestMongoIntegrationSmokeTest
     }
 
     @Test
+    public void testNonLowercaseViewName()
+    {
+        // Case insensitive schema name
+        MongoCollection<Document> collection = client.getDatabase("NonLowercaseSchema").getCollection("test_collection");
+        collection.insertOne(new Document(ImmutableMap.of("Name", "abc", "Value", 1)));
+
+        client.getDatabase("NonLowercaseSchema").createView("lowercase_view", "test_collection", ImmutableList.of());
+        assertQuery("SELECT value FROM nonlowercaseschema.lowercase_view WHERE name = 'abc'", "SELECT 1");
+
+        // Case insensitive view name
+        collection = client.getDatabase("test_database").getCollection("test_collection");
+        collection.insertOne(new Document(ImmutableMap.of("Name", "abc", "Value", 1)));
+
+        client.getDatabase("test_database").createView("NonLowercaseView", "test_collection", ImmutableList.of());
+        assertQuery("SELECT value FROM test_database.nonlowercaseview WHERE name = 'abc'", "SELECT 1");
+
+        // Case insensitive schema and view name
+        client.getDatabase("NonLowercaseSchema").createView("NonLowercaseView", "test_collection", ImmutableList.of());
+        assertQuery("SELECT value FROM nonlowercaseschema.nonlowercaseview WHERE name = 'abc'", "SELECT 1");
+
+        assertUpdate("DROP TABLE nonlowercaseschema.lowercase_view");
+        assertUpdate("DROP TABLE test_database.nonlowercaseview");
+        assertUpdate("DROP TABLE nonlowercaseschema.test_collection");
+        assertUpdate("DROP TABLE test_database.test_collection");
+        assertUpdate("DROP TABLE nonlowercaseschema.nonlowercaseview");
+    }
+
+    @Test
     public void testSelectView()
     {
         assertUpdate("CREATE TABLE test.view_base AS SELECT 'foo' _varchar", 1);

